tests/integrationv2/test_external_psk.py

```diff
@@ -0,0 +1,294 @@
+                  s2nd_params=[ '--psk', s2n_known_value_psk_sha384 ],
+                  s2nc_params=[ '--psk', s2n_known_value_psk_sha384 ],
+                  outcome=Outcome.psk_connection),
+    S2N_TEST_CASE(cipher=Ciphers.AES128_GCM_SHA256,
+                  certificate=None,
+                  s2nd_params=[ '--psk', s2n_psk_parameters_sha384, 
+                                '--psk', s2n_known_value_psk_sha256 ],
+                  s2nc_params=[ '--psk', s2n_psk_parameters_sha256, 
+                                '--psk', s2n_known_value_psk_sha256 ],
+                  outcome=Outcome.psk_connection),
+    # Note that the following test case's outcome is `full_handshake` instead of `connection_failed` 
+    # when `certificate` is set to `None` because s2nd uses a default certificate whencle a certificate is not provided.
+    S2N_TEST_CASE(cipher=Ciphers.CHACHA20_POLY1305_SHA256,
+                  certificate=None,
+                  s2nd_params=[ '--psk', s2n_psk_parameters_sha384 ],
+                  s2nc_params=[ '--psk', s2n_psk_parameters_sha256 ],
+                  outcome=Outcome.full_handshake),
+    S2N_TEST_CASE(cipher=Ciphers.AES128_GCM_SHA256,
+                  certificate=Certificates.ECDSA_256,
+                  s2nd_params=[ '--psk', s2n_psk_parameters_sha384 ],
+                  s2nc_params=[ '--psk', s2n_psk_parameters_sha256 ],
+                  outcome=Outcome.full_handshake)
+]
+
+OPENSSL_TEST_CASE = namedtuple(
+    'OPENSSL_TEST_CASE', 'cipher certificate s2n_params openssl_params outcome')
+
+OPENSSL_SERVER_S2NC_TEST_SUITE = [
+    OPENSSL_TEST_CASE(cipher=Ciphers.AES128_GCM_SHA256,
+                      certificate=None,
+                      s2n_params=[ '--psk', s2n_known_value_psk_sha256 ],
+                      openssl_params=openssl_known_value_psk_parameters,
+                      outcome=Outcome.psk_connection),
+    OPENSSL_TEST_CASE(cipher=Ciphers.CHACHA20_POLY1305_SHA256,
+                      certificate=None,
+                      s2n_params=[ '--psk', s2n_known_value_psk_sha256 ],
+                      openssl_params=openssl_known_value_psk_parameters,
+                      outcome=Outcome.psk_connection),
+    OPENSSL_TEST_CASE(cipher=Ciphers.CHACHA20_POLY1305_SHA256,
+                      certificate=Certificates.ECDSA_256,
+                      s2n_params=[ '--psk', s2n_psk_parameters_sha384,
+                                   '--psk', s2n_psk_parameters_sha256 ],
+                      openssl_params=openssl_known_value_psk_parameters,
+                      outcome=Outcome.connection_failed),
+    OPENSSL_TEST_CASE(cipher=Ciphers.AES256_GCM_SHA384,
+                      certificate=None,
+                      s2n_params=[ '--psk', s2n_psk_parameters_sha384,
+                                   '--psk', s2n_psk_parameters_sha256 ],
+                      openssl_params=openssl_known_value_psk_parameters,
+                      outcome=Outcome.connection_failed)
```

Comment:
Do you think it's a problem that you don't check why the connection failed? Or are the error messages not helpful / different?

Reply:
The error messages aren't very helpful and we assert for error code != 0 for this case.